domain/schema/secret.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package schema
+
+import (
+	"github.com/juju/juju/core/database/schema"
+)
+
+func secretSchema() schema.Patch {
+	return schema.MakePatch(`
+-- TODO: secret_entity normalizes the unit, application and model uuids.
+-- But we may lose the integrity check on these keys.
+CREATE TABLE
+    secret_entity (
+        uuid TEXT PRIMARY KEY,
+        unit_uuid TEXT,
```

Comment:
Just brain dumping here as @ycliuhw and myself were discussing this table and ways to work around the three foreign keys.

This seems to me more like business logic than DDL. The fact that a secret entity is living inside of the model database implies ownership by the model.

What we are trying to describe I think is a ref counting scenario where we want to count all of the ownership "uses" of the secret. If the ref counting goes to 0 then we clean up the secret?

Maybe a little contrived but something like this could work.....

```sql
CREATE TABLE secret_interest (
    secret_entity_uuid TEXT,
    interest_uuid TEXT,
)
```

Then on the unit, application and model there could be tables that exist called `unit_secret_interest`, `application_secret_interest` etc.

You can still write constraints that say only one thing can have a registered interest. What this design doesn't allow you to do is to easily go from a secret to the owning entity. But I don't think the current design does that either.

Hope my ramblings make some sense. Happy to flesh out my example a little more if needed.

Reply:
Here what we want is to be able to find what secrets a unit/application/model owns. 
So both define all 3 keys in the same table and 3 different tables can solve the problem. 
But the real problem is we can not access the model UUID now(or probably we don't need to know it at all).
1. put in same table;
| secret_meta_uuid | application_uuid  | unit_uuid | model_uuid |
| ------------- | ------------- | ------------- |
| a-meta-uuid | null  | an-unit-uuid  | null | 
| a-meta-uuid | null  | null | the-model-uuid(or it could be just `model` string) |

2. put in different tables;
# table - `secret_application_owner`
| secret_meta_uuid | application_uuid  |
| ------------- | ------------- |
| a-meta-uuid  | an-application-uuid  |

# table - `secret_model_owner` - no need the 2nd field at all since we know we are in the current model database.
| secret_meta_uuid |
| ------------- |
| a-meta-uuid  |

Reply:
Yeah this is what I was thinking. I don't think we need the model uuid as we are in that scope and it is a given. Introducing at something we can a foreign key onto gives us a potential point to have bad data enter the model DB. i.e a bad code path putting the wrong UUID in there.

3. Is my option for how to solve this but I think we need a secret ownership table with another id to make it all work properly. That way we can assert that only one secret uuid exists and hence only have 1 ownership claim. Bit convoluted but it get's us the end result.

Reply:
By defining the `secret_model_owned.secret_meta_uuid` as the PK also FK should solve the problem.